### DISCUSSION TITLE
How do I compile this for older Windows versions?

### DISCUSSION THREAD

**Discussion by Pokeman2003** [vhelin/wla-dx#503] [Q&A] [answered]
Okay, this is a very silly question, but I'm feeling tech illiterate at the moment because I can't figure it out. Basically, now that I'm home, I have access to my Windows 2000 computer which I normally use for development. I had expected, as one would do knowing that the concept of an assembler is very simple, that it would work on Windows 2000 since it worked just fine on modern NT versions. It did not.
The error I encountered upon trying this, if you're wondering, was something along the lines of the classic "This is not a classic Win32 application."
All I'm wondering is how to compile this for older Windows versions. Since the code _should be_ simple enough in theory, it should just work fine on something as old as Windows 3.1.

**Comment by vhelin** [comment]
Yeah, well, the code is ANSI C89 code so it can and is compiled even on an emulated late 80's Amiga so compiling it on a Windows 2000 machine should work. If you are using the Visual Studio solution file inside the repository, hmm, perhaps something special needs to be done to it? Currently it targets Windows XP machine AFAIK so you need to change some settings?

**Reply by Pokeman2003** [reply]
It's because of the whole CMake business. I don't use CMake, ever. Despite this, because you do, I feel it necessary to try to use it as well, but learning CMake's operations is making me feel incompetent(partially because I definitely am right now, this is way out of any territory I'm used to.) As for Visual Studio, it has prevented the compilation of anything pre-Vista for a long while now, and I don't know if your solution file could be opened up by an older version or not. I haven't tried that yet.

I find it kind of silly you coded it in ANSI C89, and didn't compile it for the earliest NT version possible, but ultimately it's whatever to me. You have your own workflow, I'm not about to ask you to change what works.

**Comment by Kroc** [comment]
For Win2K you could use an older version of MinGW to compile.

**Reply by Pokeman2003** [reply]
From my understanding, MingW64 is compiled to work for the earliest incarnation of Win32 and compiles to everything, so that's not the issue here. The issue is ultimately I don't understand CMake.

**Comment by vhelin** [comment]
Maybe you could modify and use some of the makefiles found in the historical folder?

Just opened the Visual Studio solution file in Visual C++ 2010 and it opened all right, but compiling didn't work, it complained about some 141_xp target/package not found. I bet if you just fix that and it'll compile just fine...

**Reply by Kroc** [reply]
"141_xp" is the XP SDK. For older versions of VC++ you would probably need to right-click on the solution and choose "retarget"; although I'm no expert on Visual Studio

**Reply by Pokeman2003** [reply]
I'll attempt it. Visual C++ 2010, that's going to be interesting to get my hands on. Haven't touched that in years, but it probably has a shot of working. Has there been new files/altered filenames since the last time it was opened?

**Reply by vhelin** [reply]
The solution file is used in CI so it can build the latest sources, it is up to date

**Reply by vhelin** [reply]
Of course if you can use newer versions of Visual Studio, they might work better... Also older ones might work. You need to try them out, I don't have Windows NT or 2000 installed...

**Reply by Pokeman2003** [reply]
Newer VStudios don't compile for anything pre-Vista, so I'm going to need something from roughly when Windows 2000 was still supported. 2010 should fit nicely into that.

**Comment by Neui** [comment, marked as answer]
[Could you test these?](https://github.com/vhelin/wla-dx/files/8673595/wla-dx-mingw32-d8b51a99ff1d8c9cb64ba4db90191718508f9c98.zip) They were compiled using MSYS2 MINGW32. The versions you would usually find are 64-bit (I guess, haven't checked) but to my knowledge Windows 2000 is 32-bit only.



I ran inside the MSys2 MinGW32 terminal:
```sh
git clone https://github.com/vhelin/wla-dx wla-dx-to-build
mkdir wla-dx-to-build/build-mingw32
cd wla-dx-to-build/build-mingw32
cmake .. -G 'MSYS Makefiles'
make -j8
strip binaries/*.exe # Optional
```

Also, there is an "XP-compatible toolchain" you can install via the Visual Studio Installer to compile for Windows XP, not sure if the resulting binaries also works for earlier versions of Windows.

**Reply by Pokeman2003** [reply]
I'll test them very soon. Windows 2000 has a 64 bit version, but it's IA64, not x86_64(I think, I could be wrong,) so it's nothing you'd want to compile for anyways(nor something I'm even bothering to run.)

**Reply by Neui** [reply]
Yes I meant x86_64, but according to [Wikipedia for Windows 2000](https://en.wikipedia.org/wiki/IA-64) there is no IA-64 support either, seems IA-64 was [released a year later](https://en.wikipedia.org/wiki/IA-64), and seems it was first supported in a [special version of XP](https://en.wikipedia.org/wiki/Windows_XP_editions#Windows_XP_64-Bit_Edition).

**Reply by Pokeman2003** [reply]
I counter with this [article](http://www.windowswiki.info/2012/07/18/the-development-of-ia-64-windows/), discussing the development of IA64, which Microsoft had directly collaborated on and had debuted a 64 bit version of Windows to help show off IA64. It doesn't specify which version of Windows, but given the year, 1999, and the market IA64 was intended for, it was likely this was a version of Windows 2000. This is probably the stray prototype Windows Advanced Server Limited Edition we've known about for years, with [the Windows Server 2000 start screen with MS spray-paint of "64" on it.](http://www.windowswiki.info/wp-content/uploads/2012/07/ia64win2000srv.png) 
Anyways, got it on my flashdrive now, will test in about 20 minutes.

**Reply by Pokeman2003** [reply]
Alright well this seems to work just fine. Thanks for compiling these, I'll keep this all in mind. If you could do it, then I think I could do it for the next version. Thanks for helping!
